### PR TITLE
Allow arbitrary input shape for convolution models

### PR DESCRIPTION
**Pull request by FloopCZ**

### CODE REVIEW

File: emloop_tensorflow/models/conv.py [hunk 1]
```diff
@@ -198,33 +198,34 @@ def cnn_autoencoder(x: tf.Tensor,
...
-            pad_bot = target_rows - pad_top - rows
-            pad_left = (target_cols - cols) // 2
-            pad_right = target_cols - pad_left - cols
-            logging.debug('Paddings: left=%d, right=%d, top=%d, bottom=%d', pad_top, pad_bot, pad_left, pad_right)
-
-            x = tf.concat([tf.zeros_like(x[:, :pad_top, :, :], dtype=x.dtype), x,
-                           tf.zeros_like(x[:, :pad_bot, :, :], dtype=x.dtype)], axis=1)
-            x = tf.concat([tf.zeros_like(x[:, :, :pad_left, :], dtype=x.dtype), x,
-                           tf.zeros_like(x[:, :, :pad_right, :], dtype=x.dtype)], axis=2)
+        rows = tf.shape(x)[1]
```
Comment (blazekadam) [inline emloop_tensorflow/models/conv.py]: you can index to `original_shape`, cant you? its just I have no idea what shenanigans are hidden underneath `tf.shape` function call...

File: emloop_tensorflow/models/conv.py [hunk 2]
```diff
@@ -198,33 +198,34 @@ def cnn_autoencoder(x: tf.Tensor,
...
+        cols = tf.shape(x)[2]
+        rows_not_divisible = tf.cast(rows % pool_product > 0, dtype=tf.int32)
```
Comment (blazekadam) [inline emloop_tensorflow/models/conv.py]: 👍

File: emloop_tensorflow/models/conv.py [hunk 3]
```diff
@@ -243,7 +244,7 @@ def cnn_autoencoder(x: tf.Tensor,
                                   skip_connections, use_bn, use_ln)
 
     logging.debug('Shape of the padded auto-encoder: %s', decoded_raw.get_shape())
-    decoded = decoded_raw[:, pad_top:-pad_bot, pad_left:-pad_right, :] if padded else decoded_raw
+    decoded = decoded_raw[:, pad_top:-pad_bot, pad_left:-pad_right, :]
```
Comment (blazekadam) [inline emloop_tensorflow/models/conv.py]: wtf? go investigate!

File: emloop_tensorflow/tests/models/conv_test.py [hunk 4]
```diff
@@ -103,22 +103,27 @@ def test_sanity_auto():
...
     fitted_encoder = ['3c3', 'mp2', '24c3', 'mp2', '48c3']  # pooling fits to the shape, no padding is required
     with tf.Graph().as_default(), tf.Session() as ses:
-        x4 = tf.ones((10, 100, 100, 3))  # 100 is not divisible by 9 -> padding will be required
+        input_x = tf.placeholder(tf.float32, shape=(None, None, None, 3))
         with tf.variable_scope('padded'):
-            _, decoded = eltf.models.cnn_autoencoder(x4, padding_encoder, use_bn=False)
+            _, decoded = eltf.models.cnn_autoencoder(input_x, padding_encoder, use_bn=False)
         with tf.variable_scope('fitted'):
-            encoded, _ = eltf.models.cnn_autoencoder(x4, fitted_encoder, use_bn=False)
+            encoded, _ = eltf.models.cnn_autoencoder(input_x, fitted_encoder, use_bn=False)
```
Comment (blazekadam) [inline emloop_tensorflow/tests/models/conv_test.py]: decoded is not tested in the fitted variant

**Comment by FloopCZ** [comment]
Thank you @blazekadam for a nice review! I addressed all the issues.

**Pull request merged**